react_juce/core/TextInputView.cpp

```diff
@@ -0,0 +1,207 @@
+    {
+        insertedAsControlledValue = true;
+        setText(value);
+    }
+
+    void TextInput::setMaxLength(int maxLen)
+    {
+        maxLength = maxLen;
+        setInputRestrictions(maxLen);
+    }
+
+    void TextInput::setPlaceholderText(const juce::String &text)
+    {
+        placeholderText = text;
+        setTextToShowWhenEmpty(placeholderText, placeholderColour);
+    }
+
+    void TextInput::setPlaceholderColour(const juce::Colour &colourToUse)
+    {
+        placeholderColour = colourToUse;
+        setTextToShowWhenEmpty(placeholderText, placeholderColour);
+    }
+
+    void TextInput::insertTextAtCaret(const juce::String &textToInsert)
+    {
+        juce::String currentValue = getText();
+        juce::TextEditor::insertTextAtCaret(textToInsert);
+        juce::String newValue = getText();
+        if (currentValue == newValue)
+        {
+            return;
+        }
+
+        // Invoke JavaScript's `input` event.
+        if (props.contains(TextInputView::onInputProp) && props[TextInputView::onInputProp].isMethod())
+        {
+            std::array<juce::var, 1> args{{detail::makeInputEventObject(newValue)}};
+            juce::var::NativeFunctionArgs nfArgs(juce::var(), args.data(), static_cast<int>(args.size()));
+            std::invoke(props[TextInputView::onInputProp].getNativeFunction(), nfArgs);
+        }
+
+        dirty = true;
+
+        if (controlled && !insertedAsControlledValue)
+        {
+            undo();
+        }
+        insertedAsControlledValue = false;
+    }
+
+    //==============================================================================
+
+    void TextInput::textEditorReturnKeyPressed(juce::TextEditor &)
+    {
+        invokeChangeEventIfNeeded();
+    }
+
+    void TextInput::textEditorFocusLost(juce::TextEditor &)
+    {
+        invokeChangeEventIfNeeded();
+    }
+
+    void TextInput::invokeChangeEventIfNeeded()
+    {
+        if (dirty)
+        {
+            // Invoke JavaScript's `change` event.
+            if (props.contains(TextInputView::onChangeProp) && props[TextInputView::onChangeProp].isMethod())
+            {
+                std::array<juce::var, 1> args{{detail::makeChangeEventObject(getText())}};
+                juce::var::NativeFunctionArgs nfArgs(juce::var(), args.data(), static_cast<int>(args.size()));
+                std::invoke(props[TextInputView::onChangeProp].getNativeFunction(), nfArgs);
+            }
+            dirty = false;
+        }
+    }
+
+    //==============================================================================
+
+    TextInputView::TextInputView()
+        : textInput(props)
+    {
+        addAndMakeVisible(textInput);
+        textInput.addListener(&textInput);
+        textInput.setPopupMenuEnabled(false);
+    }
+
+    void TextInputView::setProperty(const juce::Identifier &name, const juce::var &value)
+    {
+        View::setProperty(name, value);
```

Comment:
Could we maybe get some newlines between these if statements? Feels a bit more readable that way. Personally I'd also have a newline below the `throw` statements.

Reply:
I fixed it in 973a63c.